chiselFrontend/src/main/scala/chisel3/core/Module.scala

```diff
@@ -79,34 +71,122 @@ object Module {
   def reset: Bool = Builder.forcedReset
 }
 
-/** Abstract base class for Modules, which behave much like Verilog modules.
-  * These may contain both logic and state which are written in the Module
-  * body (constructor).
-  *
-  * @note Module instantiations must be wrapped in a Module() call.
+/** Abstract base class for Modules, an instantiable organizational unit for RTL.
   */
-abstract class Module(
-  override_clock: Option[Clock]=None, override_reset: Option[Bool]=None)
-                     (implicit moduleCompileOptions: CompileOptions)
-extends HasId {
-  // _clock and _reset can be clock and reset in these 2ary constructors
-  // once chisel2 compatibility issues are resolved
-  def this(_clock: Clock)(implicit moduleCompileOptions: CompileOptions) = this(Option(_clock), None)(moduleCompileOptions)
-  def this(_reset: Bool)(implicit moduleCompileOptions: CompileOptions)  = this(None, Option(_reset))(moduleCompileOptions)
-  def this(_clock: Clock, _reset: Bool)(implicit moduleCompileOptions: CompileOptions) = this(Option(_clock), Option(_reset))(moduleCompileOptions)
+// TODO: seal this?
+abstract class BaseModule extends HasId {
+  //
+  // Builder Internals - this tracks which Module RTL construction belongs to.
+  //
+  if (!Builder.readyForModuleConstr) {
+    throwException("Error: attempted to instantiate a Module without wrapping it in Module().")
+  }
+  readyForModuleConstr = false
 
-  // This function binds the iodef as a port in the hardware graph
-  private[chisel3] def Port[T<:Data](iodef: T): iodef.type = {
-    // Bind each element of the iodef to being a Port
-    Binding.bind(iodef, PortBinder(this), "Error: iodef")
-    iodef
+  Builder.currentModule = Some(this)
+  Builder.whenDepth = 0
+
+  //
+  // Module Construction Internals
+  //
+  protected var _closed = false
```

Comment:
Precisely, since this function is the one both setting _closed and _component anyway, there is no time outside execution of this function where _closed and _component.isDefined are different.

Multithreading isn't in play here since the current function already can't be called by multiple concurrent threads.